core/src/HIP/Kokkos_HIP_Parallel_Team.hpp

```diff
@@ -755,10 +757,34 @@ class ParallelReduce<CombinedFunctorReducerType,
                                  !m_result_ptr_host_accessible ||
                                  !std::is_same<ReducerType, InvalidType>::value;
     if (!is_empty_range || need_device_set) {
-      const int block_count =
-          UseShflReduction
-              ? std::min(m_league_size, size_type(1024 * HIPTraits::WarpSize))
-              : std::min(static_cast<int>(m_league_size), m_team_size);
+      int block_max = 0;
+      constexpr auto light_weight =
+          Kokkos::Experimental::WorkItemProperty::HintLightWeight;
+      constexpr typename Policy::work_item_property property;
+      if ((property & light_weight) == light_weight) {
+        block_max = 2097152;
+      } else {
+        block_max = 65536;
+      }
+      constexpr int preferred_block_min = 1024;
+      int block_count_tmp               = m_league_size;
+      if (block_count_tmp < preferred_block_min) {
+        // keep blocks as is, already low parallelism
+      } else if (block_count_tmp >= block_max) {
+        block_count_tmp = block_max;
+
+      } else {
+        int nwork            = m_league_size * m_team_size;
+        int items_per_thread = (nwork + block_count_tmp * m_team_size - 1) /
+                               (block_count_tmp * m_team_size);
+        if (items_per_thread < 4) {
+          int ratio = std::min(
+              (block_count_tmp + preferred_block_min - 1) / preferred_block_min,
+              (4 + items_per_thread - 1) / items_per_thread);
+          block_count_tmp /= ratio;
+        }
+      }
+      const int block_count = block_count_tmp;
```

Comment:
Did you consider writing a function that takes the policy as argument and returns that value instead of cramming this here?
As far as I can tell the only variables here (`league_size`, `team_size`, and `work_item_properties`) are determined by the policy.

Reply:
It makes sense to me to have a block count heuristic function be a private member of the TeamPolicy class, as this heuristic is specific to TeamPolicy, and the function should only be used internally. Would that work?

Reply:
I would have defined it within that ParallelReduce class specialization and move it to some other place later as needed if using it elsewhere.
I will leave it up to @Rombur

Reply:
@Rombur, I can get this started as soon as you weigh in.

Reply:
Just create a new function in the ParallelReduce class. We can always move it later if we need it somewhere else

Reply:
@IanBogle Do you have time to finish this or do you want me to take over?

Reply:
@Rombur If you've got the time, please take over. I've been swamped lately.

Reply:
Sounds good.